### ISSUE TITLE
destroy of azurerm_storage_account_network_rules fails with default_action = Deny

### ISSUE THREAD

**Issue by grisu48** [hashicorp/terraform-provider-azurerm#15245]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

* Terraform v1.1.4 on linux_amd64
* `azurerm 2.94.0`

### Affected Resource(s)

* `azurerm_storage_account_network_rules`

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    azurerm = {
      version = ">= 2.94.0"
      source = "hashicorp/azurerm"
    }
    random = {
      version = "= 3.1.0"
      source = "hashicorp/random"
    }
  }
}

provider "azurerm" {
    features {}
}

## ---- variables ----------------------------------------------------------- ##

## general variables

variable "name" {
    default = "sagittarius"
}

variable "region" {
    default = "westeurope"
}

## -------------------------------------------------------------------------- ##

resource "random_id" "service" {
    keepers = {
        name = var.name
    }
    byte_length = 8
}

## resource group

resource "azurerm_resource_group" "storage-group" {
    name     = "${var.name}-${element(random_id.service.*.hex, 0)}"
    location = var.region
}

## virtual network

resource "azurerm_virtual_network" "storage-network" {
    name                = "${azurerm_resource_group.storage-group.name}-vnet"
    address_space       = ["192.168.0.0/16"]
    location            = var.region
    resource_group_name = azurerm_resource_group.storage-group.name
}

resource "azurerm_subnet" "storage-subnet" {
    name                 = "${azurerm_resource_group.storage-group.name}-subnet"
    resource_group_name  = azurerm_resource_group.storage-group.name
    virtual_network_name = azurerm_virtual_network.storage-network.name
    address_prefixes     = ["192.168.1.0/24"]
    service_endpoints    = ["Microsoft.Storage"]
}

## storage and NFS share

resource "azurerm_storage_account" "storage-group" {
  name                      = "storage${element(random_id.service.*.hex, 0)}"
  resource_group_name       = azurerm_resource_group.storage-group.name
  location                  = azurerm_resource_group.storage-group.location
  account_tier              = "Premium"    # Required for NFS share
  account_replication_type  = "LRS"
  account_kind              = "FileStorage"
  enable_https_traffic_only = false
  allow_blob_public_access  = false
}

resource "azurerm_storage_account_network_rules" "storage-group" {
  depends_on = [azurerm_storage_share.storage-group]

  storage_account_id         = azurerm_storage_account.storage-group.id
  default_action             = "Deny" # XXX - if this is set to "Allow" then it works
  virtual_network_subnet_ids = [azurerm_subnet.storage-subnet.id]
  
  private_link_access {
    endpoint_resource_id     = azurerm_subnet.storage-subnet.id
  }
}

resource "azurerm_storage_share" "storage-group" {
  name                 = "nfsdata"
  storage_account_name = azurerm_storage_account.storage-group.name
  quota                = 100
  enabled_protocol     = "NFS"
}
```
[storage.tf.zip](https://github.com/hashicorp/terraform-provider-azurerm/files/7995891/storage.tf.zip) (attached as zip, as `.tf` is not accepted by GitHub)

When replacing `default_action` in `azurerm_storage_account_network_rules` from `"Deny"`  to `"Allow"`, `terraform apply` and `terraform destroy` work as expected.

### Debug Output

* [debug.out](https://github.com/hashicorp/terraform-provider-azurerm/files/7995949/debug.txt)

### Expected Behaviour

* `terraform destroy` deletes all resources

### Actual Behaviour

* `terraform destroy` fails with an access denied error:

```
│ Error: shares.Client#GetProperties: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailure" Message="This request is not authorized to perform this operation.\nRequestId:<REDACTED>\nTime:2022-02-03T15:29:14.6461222Z"
│ 
│   with azurerm_storage_share.storage-group,
│   on storage.tf line 88, in resource "azurerm_storage_share" "storage-group":
│   88: resource "azurerm_storage_share" "storage-group" {
│ 
╵
```

I can delete the whole resource group from the WebUI without any problems, using the same user.

### Steps to Reproduce

1. `terraform apply`
2. `terraform destroy`

As stated above, if `default_action` is set to `Allow` this issue does not occur and `terraform destroy` works as expected.

**Comment by tombuildsstuff** [comment]
hi @grisu48 

Thanks for opening this issue.

Taking a look through this appears to be the same root cause as / a duplicate of #2977 - rather than having multiple issues open tracking the same thing I'm going to close this issue in favour of that one; would you mind subscribing to #2977 for updates?

Thanks!

**Comment by grisu48** [comment]
Thank you. It wasn't obvious to me that those two originate from the same reason as this one happens in `terraform destroy`, while https://github.com/hashicorp/terraform-provider-azurerm/issues/2977 refers to the account creation phase.

Will follow-up there, sorry for the noise :+1:

**Comment by tombuildsstuff** [comment]
@grisu48 fwiw the same root cause there is that the Resource Manager and Data Plane API's unfortunately aren't feature-compatible - so we have to reach out to the Data Plane API to determine if there's a diff (as the fields don't also exist in the Resource Manager API) - which unfortunately isn't accessible when Network Rules are in effect. Fixing that is being tracked in #2977, but it's ultimately waiting on the Azure Storage team to make the same functionality available in both APIs.

**Comment by grisu48** [comment]
Thank you for the explanation. Much appreciated, you guys are doing good work! :+1:

**Comment by tspearconquest** [comment]
In case it helps, the last few comments on #2977 have a workaround you can use until the fix is released.

**Comment by grisu48** [comment]
> In case it helps, the last few comments on #2977 have a workaround you can use until the fix is released.

Yep, thanks for pointing this out. In the end this was also the workaround that we applied. More specifically the `ip_rules = [chomp(data.http.myip.body)]` block in our `azurerm_storage_account_network_rules` section:

```terraform
resource "azurerm_storage_account_network_rules" "openqa-group" {
  depends_on = [azurerm_storage_share.openqa-group]

  storage_account_id         = azurerm_storage_account.openqa-group.id
  default_action             = "Deny"
  virtual_network_subnet_ids = [azurerm_subnet.openqa-subnet.id]
  ip_rules = [chomp(data.http.myip.body)]
  
  private_link_access {
    endpoint_resource_id     = azurerm_subnet.openqa-subnet.id
  }
}
```

Not the most beautiful solution, but it seems to do the job :slightly_smiling_face:

**Comment by tspearconquest** [comment]
[This might be more elegant](https://github.com/hashicorp/terraform-provider-azurerm/issues/2977#issuecomment-1010353063) :)

**Comment by tspearconquest** [comment]
I'm assuming you are running terraform from a subnet that isn't peered to the `openqa-subnet` resource?
You shouldn't need IP rules when using a Private Endpoint if you run terraform from the subnet where the endpoint is located or a subnet peered to it.

What you need in order to do that is endpoint in a subnet that is not managed by Terraform. Have that subnet peered to the subnet where you run Terraform from, so that you can access the storage without the IP rules.
